api/db/migrations/20200515163806_add-column-createdat-updatedat-to-memberships.js

```diff
@@ -0,0 +1,22 @@
+const TABLE_NAME = 'memberships';
+const CREATED_AT_COLUMN = 'createdAt';
+const UPDATED_AT_COLUMN = 'updatedAt';
+const UNIX_EPOCH = '1970-01-01';
+
+exports.up = function(knex) {
+
+  // For actual data in table, mark as not applicable (use UNIX_EPOCH)
+  return knex.schema.table(TABLE_NAME, (table) => {
+    table.dateTime(CREATED_AT_COLUMN).notNullable().defaultTo(UNIX_EPOCH);
+    table.dateTime(UPDATED_AT_COLUMN).notNullable().defaultTo(UNIX_EPOCH);
+  });
+
+};
+
+exports.down = function(knex) {
+  return knex.schema.table(TABLE_NAME, (table) => {
+    table.dropColumn(UPDATED_AT_COLUMN);
+    table.dropColumn(CREATED_AT_COLUMN);
+  });
+};
```

Comment:
pourquoi faire 2 migrations différentes ?

Reply:
Je pensais que c'était clair dans la description de PR, mais à priori non..
> Alimentation de ces champs:
>  via l'application, par un modèle Bookshelf paramétré avec ces propriétés
>    pour les données existantes, avec la date Unix epoch (01/01/1970)
 >   pour les données manipulées via une invite SQL, avec une valeur par défaut, l'horodatage du jour

Reply:
Si tu sais comment passer 2 scripts dans une migration, on peut les fusionnner !

Reply:
C'est plus une exécution séquentielle, du coup le .all n'est pas adapté. On a préféré exprimer dans deux fichiers différents car ce sont deux intentions différentes. Peut être le faire avec un .raw et de le faire en sql, on regarde cette piste.

Reply:
@octo-pito quel est le problème que vous avez rencontré en essayant d'exécuter deux instructions de modification du schéma dans le même script de migration ? 😮

Reply:
D'ailleurs plutôt que deux `ALTER TABLE` successifs, pour arriver à ce que vous cherchez je me demande si ajouter les colonnes avec la bonne valeur par défaut (`now()`) puis faire un `UPDATE` qui réinitialise toutes les lignes à la date que vous voulez fixer, ne serait pas un peu plus propre (évite 2 modifications de schémas d'affilée). D'ailleurs c'est si fondamental que ça de mettre `01/01/1970` dans ces colonnes ? 🤔

Reply:
@jonathanperret pas de problème pour éxecuter deux scripts d'alter dans un même fichier de migration c'était plus pour exprimer une intention par script. Je viens de les mutualiser à l'occasion.
Ta proposition était notre option 1, on a préféré d'utiliser l'option d'alter pour les deux raisons suivantes:
 -Pour obtenir un verrou explicite du type ACCESS EXCLUSIVE qui va garantir qu'il y aura pas d'incohérence de la data lors d'un accès conccurentiel.
 -On pense que c'est plus rapide qu'un update sur la colonne.

On avait l'option de faire une mise à jour mais cette fois avec un SELECT FOR UPDATE qui permet d'obtenir un verrou ROW SHARE moins exclusive.

https://docs.postgresql.fr/10/explicit-locking.html

Reply:
(avertissement : le message ci-dessous est d'une longueur disproportionnée par rapport à l'enjeu de cette migration, je me suis laissé emporter, désolé 🤭)

> Ta proposition était notre option 1, on a préféré utiliser l'option d'alter pour les deux raisons suivantes:
> -Pour obtenir un verrou explicite du type ACCESS EXCLUSIVE qui va garantir qu'il y aura pas d'incohérence de la data lors d'un accès concurrentiel.

L'accès concurrentiel à cette colonne pendant la migration ne devrait pas vraiment être un problème :
 * elle est créée par cette migration ;
 * le nouveau code (susceptible d'alimenter la colonne) n'est pas mis en ligne avant que l'exécution du script de migration ne soit terminé (cf. https://doc.scalingo.com/platform/app/postdeploy-hook ).

De toute façon le verrou `ACCESS EXCLUSIVE` est acquis nécessairement lors du premier `ALTER TABLE` qui ajoute la colonne, et comme la migration est exécutée dans une transaction ce verrou reste effectif jusqu'à la fin de la migration. Donc, que la rectification des dates soit faite avec un `UPDATE` ou avec un `ALTER TABLE` supplémentaire, dans les deux cas la table aura été verrouillé exclusivement.

> -On pense que c'est plus rapide qu'un update sur la colonne.

On a de grosses tables mais `memberships` n'est pas l'une d'entre elles, avec un peu moins de 20.000 lignes. C'est vrai cela dit qu'une simple modification de méta-données comme un changement de valeur par défaut prendra toujours moins de temps que la mise à jour de toutes les lignes.

> Avec l'option update, il faut le faire via SELECT FOR UPDATE qui permet d'obtenir un verrou ROW SHARE moins exclusive.

Pas besoin, la table est déjà verrouillée, cf. plus haut. Et quoi qu'il en soit, `SELECT FOR UPDATE` n'est utile que si on a un `SELECT` suivi d'un `UPDATE` (et qu'on veut garantir l'atomicité de l'ensemble) ; en l'occurrence il s'agissait d'un unique `UPDATE` qui est naturellement atomique.

> La date 01/01/1970 est utilisé pour remplir la colonne vu qu'on a rajouté la contrainte de non nullité et de distinguer les liens non tracés.

Je comprends l'intention, et du coup je m'interroge : cette date particulière servirait donc à communiquer le fait qu'on ne dispose pas de l'information correspondante. C'est effectivement une technique répandue, quand une structure de données contient une case pour une valeur et pas de case pour indiquer si la valeur est présente ou non, que de désigner une valeur particulière du type comme signifiant l'absence de donnée valide. Par exemple, la fonction système [open()](http://man7.org/linux/man-pages/man2/open.2.html#RETURN_VALUE) renvoie un descripteur de fichier sous la forme d'un entier, ou `-1` si l'appel a échoué. Mais en l'occurrence, on se trouve dans un système de types plus riche que (par exemple) les types primitifs de C, où il existe un mécanisme universel pour étendre un type en ajoutant une valeur spéciale représentant l'absence de donnée, à savoir `NULL`. Conceptuellement, le type `int NULL` possède exactement un habitant de plus que le type `int NOT NULL` ce qui permet d'éviter d'avoir à "sacrifier" une valeur possible de `int`.

Je trouve dommage du coup de restreindre la colonne à un type non étendu (`datetime NOT NULL`) pour ensuite désigner une valeur spécifique de ce type pour jouer le rôle de `NULL`… au lieu de simplement donner à la colonne le type approprié compte tenu du fait que vous tenez à distinguer les lignes n'ayant pas de donnée valide pour ce champ, à savoir `datetime NULL`. Ceci communiquerait automatiquement à tous les utilisateurs de cette table, d'une façon standard, le fait que la donnée n'est pas disponible — au lieu de devoir coder spécifiquement une reconnaissance d'une date en particulier (01/01/1970) dans le code qui voudrait "distinguer les liens non tracés".

Ou alternativement, de conserver le type `NOT NULL` et admettre que la date de la migration ne soit pas une si mauvaise "vraie" date à stocker dans ce champ pour les lignes préexistantes (elle est plus correcte que 01/01/1970 de 50 ans environ 😄).

Reply:
J'ai appris pleins de choses merci !

Reply:
Merci pour ton retour et tes éclaircissements. J'ai pris en compte ta suggestion

Reply:
Je laisse cette conversation non résolu pour le partage :)